Content.Client/ContextMenu/UI/EntityMenuPresenter.cs

```diff
@@ -0,0 +1,421 @@
+using System.Collections.Generic;
+using System.Linq;
+using Content.Client.Examine;
+using Content.Client.Items.Managers;
+using Content.Client.Verbs;
+using Content.Client.Viewport;
+using Content.Shared.CCVar;
+using Content.Shared.Input;
+using Content.Shared.Interaction.Helpers;
+using Robust.Client.GameObjects;
+using Robust.Client.Graphics;
+using Robust.Client.Input;
+using Robust.Client.Player;
+using Robust.Client.State;
+using Robust.Client.UserInterface;
+using Robust.Client.UserInterface.Controls;
+using Robust.Shared.Configuration;
+using Robust.Shared.Containers;
+using Robust.Shared.GameObjects;
+using Robust.Shared.Input;
+using Robust.Shared.Input.Binding;
+using Robust.Shared.IoC;
+using Robust.Shared.Log;
+using Robust.Shared.Maths;
+using Robust.Shared.Timing;
+namespace Content.Client.ContextMenu.UI
+{
+    /// <summary>
+    ///     This class handles the displaying of the entity context menu.
+    /// </summary>
+    /// <remarks>
+    ///     In addition to the normal <see cref="ContextMenuPresenter"/> functionality, this also provides functions get
+    ///     a list of entities near the mouse position, add them to the context menu grouped by prototypes, and remove
+    ///     them from the menu as they move out of sight.
+    /// </remarks>
+    public sealed partial class EntityMenuPresenter : ContextMenuPresenter
+    {
+        [Dependency] private readonly IEntitySystemManager _systemManager = default!;
+        [Dependency] private readonly IItemSlotManager _itemSlotManager = default!;
+        [Dependency] private readonly IEntityManager _entityManager = default!;
+        [Dependency] private readonly IPlayerManager _playerManager = default!;
+        [Dependency] private readonly IStateManager _stateManager = default!;
+        [Dependency] private readonly IInputManager _inputManager = default!;
+        [Dependency] private readonly IConfigurationManager _cfg = default!;
+        [Dependency] private readonly IGameTiming _gameTiming = default!;
+        [Dependency] private readonly IUserInterfaceManager _userInterfaceManager = default!;
+        [Dependency] private readonly IEyeManager _eyeManager = default!;
+
+        private readonly VerbSystem _verbSystem;
+
+        /// <summary>
+        ///     This maps the currently displayed entities to the actual GUI elements.
+        /// </summary>
+        /// <remarks>
+        ///     This is used remove GUI elements when the entities are deleted. or leave the LOS.
+        /// </remarks>
+        public Dictionary<IEntity, EntityMenuElement> Elements = new();
+
+        public EntityMenuPresenter(VerbSystem verbSystem) : base()
+        {
+            IoCManager.InjectDependencies(this);
+
+            _verbSystem = verbSystem;
+
+            _cfg.OnValueChanged(CCVars.ContextMenuGroupingType, OnGroupingContextMenuChanged, true);
+
+            CommandBinds.Builder
```

Comment:
shouldnt the commandbinds stuff be handled entirely by ContextMenuPresenter and not its children?

Reply:
There were some instances here where I forgot to update ContextMenu -> EntityMenu, which possibly caused some confusion. I've fixed that so `HandleOpenContextMenu` is now `HandleOpenEntityMenu` and the `ContextMenuGroupingType` cvar is now `EntityMenuGroupingType`.

This keybind is specifically for opening the entity menu, and I don't think there is an easy way to move it into `ContextMenuPresenter` because it exists to open a specific context menu type. The only way I could think of that working would require the keybinding to call a generic `HandleOpenContextMenu()` that has to be overridden by each type of context menu, which then checks whether the entity/location that was right clicked is appropriate to that menu type?

But I think just keeping it how it is is probably probably fine. Though maybe the keybinding itself should be moved into `VerbSystem` instead of `EntityMenuPresenter`.